### PR TITLE
[QNN] Add hardswish int8 impl using table lookup

### PR DESCRIPTION
**Pull request by zhaoyang-star**
Registered the hardswish unary elementwise op

@AndrewZhaoLuo @mbrookhart

### CODE REVIEW

**Comment by AndrewZhaoLuo** [comment]
I'll take a look today or tomorrow

File: python/tvm/relay/frontend/qnn_torch.py [hunk 1]
```diff
@@ -987,6 +983,21 @@ def hardsigmoid(x):
             dequantized_hswish, output_scale, output_zero_point, out_dtype="uint8"
         )
 
+    def _impl_int8(inputs):
+        output_scale = _expr.const(inputs[1])
+        output_zero_point = _expr.const(inputs[2])
+        input_scale = _expr.const(inputs[3])
+        input_zero_point = _expr.const(inputs[4])
+        return relay.qnn.op.hardswish(
+            input[0], input_scale, input_zero_point, output_scale, output_zero_point
```
Comment (AndrewZhaoLuo) [inline python/tvm/relay/frontend/qnn_torch.py:992]: this should be `inputs[0]` not `input[0]`

**Review by AndrewZhaoLuo — APPROVED** [review]
LGTM, the hexagon was probably just flaky so you can rerun CI with an empty commit:

`git commit -m 'jostle ci' --allow-empty` and `git push`

As for the other errors, see the comment

**Comment by zhaoyang-star** [comment]
> LGTM, the hexagon was probably just flaky so you can rerun CI with an empty commit:
> 
> `git commit -m 'jostle ci' --allow-empty` and `git push`
> 
> As for the other errors, see the comment

Thanks for your kind help. I have fix the typo and recommit with `git commit -m 'jostle ci' --allow-empty`. But tasks for Hexagon still failed. The error log:
```
LLVM ERROR: Cannot select: 0x451c698: i32 = fp_to_fp16 0x44eba68
```
It seems ci doesn't skip tasks for Hexagon?

**Comment by AndrewZhaoLuo** [comment]
I'm not sure, maybe @mehrdadh or @cconvey knows something? 

Maybe you just need to rebase on main and try again? Not sure

**Comment by mehrdadh** [comment]
@zhaoyang-star I checked our CI on TVM main and I don't see this error. Can you reproduce the error locally using ci_hexagon docker image. It might be related to your PR

**Comment by zhaoyang-star** [comment]
> LLVM ERROR: Cannot select: 0x451c698: i32 = fp_to_fp16 0x44eba68

@mehrdadh Thanks for your quick reply. Maybe it is my PR that cause the failure. @AndrewZhaoLuo 
I will double check PR and try to reproduce the error locally.

**Comment by zhaoyang-star** [comment]
> @mehrdadh Thanks for your quick reply. Maybe it is my PR that cause the failure. @AndrewZhaoLuo
> I will double check PR and try to reproduce the error locally.

**Comment by AndrewZhaoLuo** [comment]
Nah, it's just flaky -- I think there was some planned work with the CI machines so you got hit with the outage. Just jostle one more time and let's get this merged!

**Comment by AndrewZhaoLuo** [comment]
Also apparently you can do this:

**Comment by AndrewZhaoLuo** [comment]
@tvm-bot rerun

**Comment by zhaoyang-star** [comment]
All ci tasks have passed
@AndrewZhaoLuo please merge if no more opinion ^_^

**Comment by AndrewZhaoLuo** [comment]
Hmm, do you have a profiler report?  

I am curious since I would expect runtimes to be better vs dq - fp32 - q. Do you have a repo to reproduce?

**Comment by zhaoyang-star** [comment]
> Hmm, do you have a profiler report?
> 
> I am curious since I would expect runtimes to be better vs dq - fp32 - q. Do you have a repo to reproduce?

I replaced all relu with hswish in resnet50.  

- only use one cpu core
- benchmark int8 model

 Quantized Model  | Inference Time(msec)
---|---
resnet50(relu) | 1149
resnet50(hswish) w/o LUT | 1210
resnet50(hswish, LUT) | 1171

About 3% speedup by using LUT. I also tried yolov5 with hswish model, which is about 9% speedup by LUT.

**Comment by AndrewZhaoLuo** [comment]
Hmm yeah this makes sense, i would expect LUT to be slower than ReLU as it requires more memory access. 

I suspect perhaps the activation functions just don't take much time? Really ReLU is close to the fastest you can go probably. You can maybe see the upper value for speedup by removing all activations.

Still technically a little bit of improvement!

**Comment by zhaoyang-star** [comment]
> > Hmm, do you have a profiler report?
> > I am curious since I would expect runtimes to be better vs dq - fp32 - q. Do you have a repo to reproduce?
> 
> Based on `tests/python/fronend/pytorch/test_fx_quant.py`, I replaced all relu with hswish in resnet50.
> 
> * only use one cpu core
> * benchmark int8 model
> 
> Quantized Model	Inference Time(msec)
> resnet50(relu)	1149
> resnet50(hswish) w/o LUT	1210
> resnet50(hswish, LUT)	1171
> About 3% speedup by using LUT. I also tried yolov5 with hswish model, which is about 9% speedup by LUT.

Maybe there is something wrong when I created the resnet50 with hswish.
I used a quantized YOLOv5s in which has hswish. The perf improved 50.2% ^_^ 

 Quantized Model  | Inference Time(msec)
---|---
YOLOv5s(hswish) w/o LUT | 18.88
YOLOv5s(hswish, LUT) | 12.57

**Pull request merged**